File: stack-python/app/tools/get_approval_detail.py
from typing import Any

from pydantic_ai import RunContext

from app.policy import authorize

from .common import error, visible_detail
from .context import ToolContext, resolve_context


async def get_approval_detail(ctx: RunContext[ToolContext], timesheet_id: str) -> dict[str, Any]:
    """Fetch one timesheet by id with every field, for review before a decision. Returns not_found if the id does not exist or the current user is not allowed to see it."""
    tool_ctx = resolve_context(ctx)
    verdict = authorize(tool_ctx.policy_caller, "read", None)
    if not verdict.allowed:
        return error(verdict.error or "not_authorized", verdict.message)
    row = await visible_detail(tool_ctx, timesheet_id)
    return row.model_dump(by_alias=True) if row else error("not_found", "Timesheet was not found or is outside caller coverage.")

File: stack-typescript/src/tools/get-approval-detail.ts
import { tool } from "ai";
import { z } from "zod";
import { authorize, tenantFor } from "../policy/approval.js";
import type { ToolError } from "../types.js";
import { upstreamFor } from "../upstream/index.js";
import { callerFromOptions, callerSchema } from "./context.js";

export const getApprovalDetail = tool({
  description: "Fetch one timesheet by id with every field, for review before a decision. Returns not_found if the id does not exist or the current user is not allowed to see it.",
  inputSchema: z.object({ timesheet_id: z.string() }).strict(),
  contextSchema: callerSchema,
  execute: async ({ timesheet_id }, options) => {
    const caller = callerFromOptions(options);
    const baseline = authorize(caller, "read", null);
    if (!baseline.allowed) return baseline;
    const tenant = tenantFor(caller);
    const row = await upstreamFor(caller).detail(tenant, timesheet_id);
    if (!row) return { error: "not_found", message: "Timesheet was not found or is outside caller coverage." } satisfies ToolError;
    const verdict = authorize(caller, "read", row);
    return verdict.allowed ? row : { error: "not_found", message: "Timesheet was not found or is outside caller coverage." } satisfies ToolError;
  },
});
